File: test_complete_user_flow.js
#!/usr/bin/env node

/**
 * Comprehensive User Flow Test
 * Tests the complete user journey: Registration → Login → Onboarding → Dashboard
 * This will help identify exactly where the onboarding loop issue occurs.
 */

const axios = require('axios');

const API_BASE = 'http://localhost:8000';
const FRONTEND_BASE = 'http://localhost:3000';

// Test configuration
const TEST_USER = {
  email: `test_${Date.now()}@example.com`,
  password: 'TestPass123!',
  first_name: 'Test',
  last_name: 'User',
  phone: '+1234567890'
};

console.log('🧪 COMPREHENSIVE USER FLOW TEST');
console.log('================================');
console.log(`Testing with user: ${TEST_USER.email}`);
console.log('');

async function waitForServices() {
  console.log('🔍 Checking if services are running...');
  
  // Check backend
  try {
    const response = await axios.get(`${API_BASE}/docs`, { timeout: 5000 });
    console.log('✅ Backend is running');
  } catch (error) {
    console.log('❌ Backend not responding. Please start: cd backend && uvicorn app.main:app --reload');
    process.exit(1);
  }
  
  // Check frontend
  try {
    const response = await axios.get(`${FRONTEND_BASE}`, { timeout: 5000 });
    console.log('✅ Frontend is running');
  } catch (error) {
    console.log('❌ Frontend not responding. Please start: cd frontend && npm run dev');
    process.exit(1);
  }
  
  console.log('');
}

async function testRegistration() {
  console.log('1️⃣ TESTING USER REGISTRATION');
  console.log('----------------------------');
  
  try {
    const response = await axios.post(`${API_BASE}/api/v1/auth/register`, {
      email: TEST_USER.email,
      password: TEST_USER.password,
      first_name: TEST_USER.first_name,
      last_name: TEST_USER.last_name,
      phone: TEST_USER.phone
    });
    
    console.log('✅ Registration successful');
    console.log(`   User ID: ${response.data.user?.id || 'Not provided'}`);
    console.log(`   Email: ${response.data.user?.email}`);
    console.log(`   Onboarding completed: ${response.data.user?.onboarding_completed || false}`);
    console.log(`   Onboarding step: ${response.data.user?.onboarding_step || 1}`);
    
    return response.data;
  } catch (error) {
    console.log('❌ Registration failed');
    console.log(`   Error: ${error.response?.data?.detail || error.message}`);
    throw error;
  }
}

async function testLogin() {
  console.log('\n2️⃣ TESTING USER LOGIN');
  console.log('---------------------');
  
  try {
    const response = await axios.post(`${API_BASE}/api/v1/auth/login`, {
      email: TEST_USER.email,
      password: TEST_USER.password
    });
    
    console.log('✅ Login successful');
    console.log(`   Access token: ${response.data.access_token ? 'Provided' : 'Missing'}`);
    console.log(`   User ID: ${response.data.user?.id}`);
    console.log(`   Onboarding completed: ${response.data.user?.onboarding_completed || false}`);
    console.log(`   Onboarding step: ${response.data.user?.onboarding_step || 1}`);
    
    return {
      token: response.data.access_token,
      user: response.data.user
    };
  } catch (error) {
    console.log('❌ Login failed');
    console.log(`   Error: ${error.response?.data?.detail || error.message}`);
    throw error;
  }
}

async function testOnboardingFlow(token, userId) {
  console.log('\n3️⃣ TESTING ONBOARDING FLOW');
  console.log('---------------------------');
  
  const headers = { Authorization: `Bearer ${token}` };
  
  // Step 1: Check initial onboarding state
  console.log('\n📋 Checking initial onboarding state...');
  try {
    const response = await axios.get(`${API_BASE}/api/v1/onboarding/${userId}`, { headers });
    console.log(`   Current step: ${response.data.step_number}`);
    console.log(`   Step data: ${JSON.stringify(response.data.data)}`);
  } catch (error) {
    console.log(`   ⚠️ Could not get onboarding state: ${error.response?.data?.detail || error.message}`);
  }
  
  // Step 2: Complete onboarding steps
  const onboardingSteps = [
    {
      step: 1,
      name: 'Personal Info',
      data: {
        firstName: TEST_USER.first_name,
        lastName: TEST_USER.last_name,
        phone: TEST_USER.phone
      }
    },
    {
      step: 2,
      name: 'Company Details',
      data: {
        company: 'Test Real Estate',
        position: 'Test Agent',
        licenseNumber: 'TEST123'
      }
    },
    {
      step: 3,
      name: 'AI Preferences',
      data: {
        aiStyle: 'Professional',
        aiTone: 'Friendly'
      }
    },
    {
      step: 4,
      name: 'Social (Skip)',
      data: {}
    },
    {
      step: 5,
      name: 'Terms',
      data: {
        termsAccepted: true,
        privacyAccepted: true
      }
    }
  ];
  
  for (const stepInfo of onboardingSteps) {
    console.log(`\n📝 Completing step ${stepInfo.step}: ${stepInfo.name}`);
    try {
      const response = await axios.post(`${API_BASE}/api/v1/onboarding/${userId}`, {
        step_number: stepInfo.step,
        data: stepInfo.data
      }, { headers });
      
      console.log(`   ✅ Step ${stepInfo.step} completed successfully`);
    } catch (error) {
      console.log(`   ❌ Step ${stepInfo.step} failed: ${error.response?.data?.detail || error.message}`);
      throw error;
    }
  }
  
  // Step 3: Complete onboarding
  console.log('\n🎯 Completing onboarding...');
  try {
    const response = await axios.post(`${API_BASE}/api/v1/onboarding/${userId}/complete`, {}, { headers });
    console.log('   ✅ Onboarding completion API call successful');
    console.log(`   Response: ${JSON.stringify(response.data)}`);
  } catch (error) {
    console.log(`   ❌ Onboarding completion failed: ${error.response?.data?.detail || error.message}`);
    throw error;
  }
  
  // Step 4: Verify user state after completion
  console.log('\n🔍 Verifying user state after completion...');
  try {
    const response = await axios.get(`${API_BASE}/api/v1/auth/me`, { headers });
    console.log(`   Onboarding completed: ${response.data.onboarding_completed}`);
    console.log(`   Onboarding step: ${response.data.onboarding_step}`);
    console.log(`   User ID: ${response.data.id}`);
    
    // This is the critical check - if onboarding_completed is false, that's the bug!
    if (!response.data.onboarding_completed) {
      console.log('   🚨 BUG FOUND: onboarding_completed is still false after completion!');
      return { bugFound: true, userState: response.data };
    } else {
      console.log('   ✅ User correctly marked as onboarding completed');
      return { bugFound: false, userState: response.data };
    }
  } catch (error) {
    console.log(`   ❌ Could not verify user state: ${error.response?.data?.detail || error.message}`);
    throw error;
  }
}

async function testPropertyFeatures(token, userId) {
  console.log('\n4️⃣ TESTING PROPERTY FEATURES');
  console.log('-----------------------------');
  
  const headers = { Authorization: `Bearer ${token}` };
  
  // Test property creation
  console.log('\n🏠 Testing property creation...');
  try {
    const propertyData = {
      title: 'Test Property',
      description: 'A beautiful test property',
      price: 500000,
      property_type: 'House',
      bedrooms: 3,
      bathrooms: 2,
      area: 1500,
      address: '123 Test Street, Test City',
      status: 'for-sale'
    };
    
    const response = await axios.post(`${API_BASE}/api/v1/properties`, propertyData, { headers });
    console.log('   ✅ Property creation successful');
    console.log(`   Property ID: ${response.data.id}`);
    return response.data.id;
  } catch (error) {
    console.log(`   ⚠️ Property creation not available: ${error.response?.data?.detail || error.message}`);
    return null;
  }
}

async function runCompleteTest() {
  try {
    await waitForServices();
    
    // Test registration
    const registrationResult = await testRegistration();
    
    // Test login
    const loginResult = await testLogin();
    const { token, user } = loginResult;
    
    if (!user?.id) {
      throw new Error('User ID not provided in login response');
    }
    
    // Test onboarding flow
    const onboardingResult = await testOnboardingFlow(token, user.id);
    
    // Test property features
    await testPropertyFeatures(token, user.id);
    
    // Final summary
    console.log('\n🎉 TEST SUMMARY');
    console.log('===============');
    console.log('✅ Registration: PASSED');
    console.log('✅ Login: PASSED');
    
    if (onboardingResult.bugFound) {
      console.log('❌ Onboarding: BUG FOUND - User not marked as completed');
      console.log('\n🐛 ROOT CAUSE ANALYSIS:');
      console.log('The onboarding completion API call succeeds, but the user\'s onboarding_completed flag remains false.');
      console.log('This causes the dashboard to redirect back to onboarding, creating the loop you experienced.');
      console.log('\n🔧 RECOMMENDED FIXES:');
      console.log('1. Check the onboarding completion service in backend/app/services/onboarding_service.py');
      console.log('2. Verify the database update is actually persisting');
      console.log('3. Check if the auth/me endpoint returns updated user data');
    } else {
      console.log('✅ Onboarding: PASSED');
    }
    
    console.log('\n📊 Property features: Tested (may need API implementation)');
    
  } catch (error) {
    console.log('\n❌ TEST FAILED');
    console.log(`Error: ${error.message}`);
    console.log('\nPlease check the error details above and ensure both services are running.');
  }
}

// Run the test
runCompleteTest();

File: test_complete_user_flow.py
#!/usr/bin/env python3
"""
Complete User Flow Test
======================
Tests the complete flow: Register user -> Login -> Add properties -> Check dashboard -> Check public website
"""

import requests
import json
import sys
import time
from typing import Dict, Any, Optional

BASE_URL = "http://localhost:8000"

class UserFlowTester:
    def __init__(self):
        self.session = requests.Session()
        self.access_token = None
        self.user_id = None
        self.agent_slug = None
        self.property_id = None
        
    def log(self, message: str, status: str = "INFO"):
        """Log a message with timestamp"""
        timestamp = time.strftime("%H:%M:%S")
        print(f"[{timestamp}] {status}: {message}")
    
    def test_endpoint(self, endpoint: str, method: str = "GET", data: Dict[str, Any] = None, 
                     expected_status: int = 200, auth_required: bool = False) -> Optional[Dict]:
        """Test a single endpoint and return response data"""
        url = f"{BASE_URL}{endpoint}"
        headers = {"Content-Type": "application/json"}
        
        if auth_required and self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"
        
        try:
            if method == "GET":
                response = self.session.get(url, headers=headers)
            elif method == "POST":
                response = self.session.post(url, json=data, headers=headers)
            elif method == "PUT":
                response = self.session.put(url, json=data, headers=headers)
            else:
                self.log(f"Unsupported method: {method}", "ERROR")
                return None
            
            if response.status_code == expected_status:
                self.log(f"✅ {method} {endpoint} - Status: {response.status_code}")
                try:
                    return response.json()
                except:
                    return {"status": "success", "text": response.text}
            else:
                self.log(f"❌ {method} {endpoint} - Expected: {expected_status}, Got: {response.status_code}", "ERROR")
                self.log(f"   Response: {response.text[:200]}...", "ERROR")
                return None
                
        except Exception as e:
            self.log(f"❌ {method} {endpoint} - Error: {e}", "ERROR")
            return None
    
    def step1_register_user(self) -> bool:
        """Step 1: Register a new user"""
        self.log("=" * 60)
        self.log("STEP 1: Registering new user", "STEP")
        self.log("=" * 60)
        
        user_data = {
            "email": "testuser@example.com",
            "password": "MySecure@Pass9",
            "first_name": "Test",
            "last_name": "User",
            "phone": "+1 (555) 123-4567"
        }
        
        result = self.test_endpoint("/api/v1/auth/register", "POST", user_data, 201)
        if result:
            self.log(f"User registered successfully: {result.get('email', 'Unknown')}")
            return True
        return False
    
    def step2_login_user(self) -> bool:
        """Step 2: Login the user"""
        self.log("=" * 60)
        self.log("STEP 2: Logging in user", "STEP")
        self.log("=" * 60)
        
        login_data = {
            "email": "testuser@example.com",
            "password": "MySecure@Pass9"
        }
        
        result = self.test_endpoint("/api/v1/auth/login", "POST", login_data, 200)
        if result and "access_token" in result:
            self.access_token = result["access_token"]
            self.user_id = result.get("user", {}).get("id")
            self.log(f"User logged in successfully. Token: {self.access_token[:20]}...")
            return True
        return False
    
    def step3_check_user_profile(self) -> bool:
        """Step 3: Check user profile"""
        self.log("=" * 60)
        self.log("STEP 3: Checking user profile", "STEP")
        self.log("=" * 60)
        
        result = self.test_endpoint("/api/v1/auth/me", "GET", auth_required=True)
        if result:
            self.log(f"User profile retrieved: {result.get('email', 'Unknown')}")
            return True
        return False
    
    def step4_create_agent_profile(self) -> bool:
        """Step 4: Create agent profile"""
        self.log("=" * 60)
        self.log("STEP 4: Creating agent profile", "STEP")
        self.log("=" * 60)
        
        agent_data = {
            "agent_name": "Test Agent",
            "bio": "Experienced real estate professional with 5+ years in the industry. Specializing in residential properties.",
            "phone": "+1 (555) 987-6543",
            "email": "testagent@example.com",
            "office_address": "456 Real Estate Ave, Test City, TC 12345",
            "specialties": ["Residential", "First-time Buyers"],
            "experience": "5+ years in real estate, Licensed Realtor",
            "languages": ["English"],
            "is_public": True
        }
        
        result = self.test_endpoint("/api/v1/agent-public/create-profile", "POST", agent_data, 201, auth_required=True)
        if result:
            self.agent_slug = result.get("slug", "test-agent")
            self.log(f"Agent profile created successfully. Slug: {self.agent_slug}")
            return True
        return False
    
    def step5_add_property(self) -> bool:
        """Step 5: Add a property"""
        self.log("=" * 60)
        self.log("STEP 5: Adding property", "STEP")
        self.log("=" * 60)
        
        property_data = {
            "title": "Beautiful 2BR Condo in Downtown",
            "description": "Modern 2-bedroom condo with stunning city views. Perfect for young professionals.",
            "price": 350000,
            "property_type": "Apartment",
            "bedrooms": 2,
            "bathrooms": 2,
            "area": 900,
            "location": "Downtown Test City, TC 12345",
            "features": ["City View", "Modern Kitchen", "Parking", "Gym"],
            "is_public": True
        }
        
        result = self.test_endpoint("/api/v1/properties/", "POST", property_data, 201, auth_required=True)
        if result:
            self.property_id = result.get("id")
            self.log(f"Property added successfully. ID: {self.property_id}")
            return True
        return False
    
    def step6_check_dashboard(self) -> bool:
        """Step 6: Check dashboard"""
        self.log("=" * 60)
        self.log("STEP 6: Checking dashboard", "STEP")
        self.log("=" * 60)
        
        # Check dashboard stats
        result = self.test_endpoint("/api/v1/dashboard/stats", "GET")
        if result:
            self.log(f"Dashboard stats retrieved: {result}")
        
        # Check user properties
        result = self.test_endpoint("/api/v1/properties/", "GET", auth_required=True)
        if result:
            properties = result.get("properties", [])
            self.log(f"User has {len(properties)} properties")
            return True
        return False
    
    def step7_check_public_website(self) -> bool:
        """Step 7: Check public website"""
        self.log("=" * 60)
        self.log("STEP 7: Checking public website", "STEP")
        self.log("=" * 60)
        
        # Check agent public profile
        if self.agent_slug:
            result = self.test_endpoint(f"/api/v1/agent-public/{self.agent_slug}")
            if result:
                self.log(f"Agent public profile accessible: {result.get('agent_name', 'Unknown')}")
            else:
                # Try with mock agent
                result = self.test_endpoint("/api/v1/agent-public/john-doe")
                if result:
                    self.log("Using mock agent profile for testing")
                    self.agent_slug = "john-doe"
        
        # Check agent properties
        if self.agent_slug:
            result = self.test_endpoint(f"/api/v1/agent-public/{self.agent_slug}/properties")
            if result:
                properties = result.get("properties", [])
                self.log(f"Agent has {len(properties)} public properties")
        
        # Check agent about page
        if self.agent_slug:
            result = self.test_endpoint(f"/api/v1/agent-public/{self.agent_slug}/about")
            if result:
                self.log(f"Agent about page accessible: {result.get('agent_name', 'Unknown')}")
        
        # Test contact inquiry
        if self.agent_slug:
            inquiry_data = {
                "name": "Public Website Visitor",
                "email": "visitor@example.com",
                "phone": "+1 (555) 111-2222",
                "message": "I found your website and I'm interested in your properties!",
                "inquiry_type": "general_inquiry"
            }
            result = self.test_endpoint(f"/api/v1/agent-public/{self.agent_slug}/contact", "POST", inquiry_data)
            if result:
                self.log("Contact inquiry submitted successfully")
        
        return True
    
    def run_complete_flow(self) -> bool:
        """Run the complete user flow"""
        self.log("🚀 Starting Complete User Flow Test", "START")
        self.log("=" * 60)
        
        steps = [
            ("Register User", self.step1_register_user),
            ("Login User", self.step2_login_user),
            ("Check User Profile", self.step3_check_user_profile),
            ("Create Agent Profile", self.step4_create_agent_profile),
            ("Add Property", self.step5_add_property),
            ("Check Dashboard", self.step6_check_dashboard),
            ("Check Public Website", self.step7_check_public_website),
        ]
        
        passed = 0
        total = len(steps)
        
        for step_name, step_func in steps:
            try:
                if step_func():
                    passed += 1
                    self.log(f"✅ {step_name} completed successfully", "SUCCESS")
                else:
                    self.log(f"❌ {step_name} failed", "ERROR")
            except Exception as e:
                self.log(f"❌ {step_name} failed with exception: {e}", "ERROR")
        
        self.log("=" * 60)
        self.log(f"📊 Flow Results: {passed}/{total} steps completed successfully", "RESULT")
        
        if passed == total:
            self.log("🎉 Complete user flow test passed! All functionality is working.", "SUCCESS")
            return True
        else:
            self.log("⚠️ Some steps failed. Check the output above for details.", "WARNING")
            return False

def main():
    """Main function"""
    print("🚀 Complete User Flow Test Suite")
    print("=" * 60)
    
    # Check if server is running
    try:
        response = requests.get(f"{BASE_URL}/health", timeout=5)
        if response.status_code != 200:
            print("❌ Server is not responding properly")
            sys.exit(1)
    except Exception as e:
        print(f"❌ Cannot connect to server at {BASE_URL}: {e}")
        print("   Make sure the server is running with: python -m uvicorn app.main:app --host 0.0.0.0 --port 8000")
        sys.exit(1)
    
    print("✅ Server is running and responding")
    
    # Run the complete flow
    tester = UserFlowTester()
    success = tester.run_complete_flow()
    
    return 0 if success else 1

if __name__ == "__main__":
    sys.exit(main())